### ISSUE TITLE
websocketpp client using tlsv12_client with multiple dropped pings hangs on close

### ISSUE THREAD

**Issue by MikaelSmith** [zaphoyd/websocketpp#892]
I ran into a strange case that I've managed to reproduce with a small test client. When the error occurs, the server gets stuck in a loop busywaiting on epoll using 100% CPU in a single thread. Dev logs repeat
```
[2020-05-15 22:01:32] [devel] p = 0 bytes transferred = 0
[2020-05-15 22:01:32] [devel] asio async_read_at_least: 1
[2020-05-15 22:01:32] [devel] asio con handle_async_read
```
forever. The connection appears to have tried to terminate, but it's not clear it's in a terminated state since handle_async_read doesn't fail. It's not clear if this is a bug in websocketpp or boost.asio.

The failure scenario below is artificial, but I believe it reflects behavior I've seen in real-life with spotty networking or misbehaving load balancers where we receive several pings that seemed lost all at once right as we've decided to close the connection and attempt to re-establish it with a new server.

I'm continuing to look into how to fix it, but sharing this example out in case anyone else has encountered something similar.

----

I've reproduced it with this test client:
```
#include <websocketpp/config/debug_asio.hpp>
#include <websocketpp/client.hpp>

#include <iostream>
#include <chrono>

using websocketpp::lib::placeholders::_1;
using websocketpp::lib::placeholders::_2;
using websocketpp::lib::bind;

class debug_client {
public:
    typedef debug_client type;

    debug_client () {
        m_endpoint.set_access_channels(websocketpp::log::alevel::all);
        m_endpoint.set_error_channels(websocketpp::log::elevel::all);

        m_endpoint.init_asio();

        m_endpoint.set_tls_init_handler(bind(&type::on_tls_init,this,::_1));
        m_endpoint.set_pong_timeout(1000);
        m_endpoint.set_pong_timeout_handler(bind(&type::on_pong_timeout,this,::_1,::_2));
        m_endpoint.set_open_handler(bind(&type::on_open,this,::_1));
    }

    void start(std::string uri) {
        websocketpp::lib::error_code ec;
        auto con = m_endpoint.get_connection(uri, ec);
        if (ec) throw ec;

        m_endpoint.connect(con);
        m_endpoint.run();
    }

    websocketpp::lib::shared_ptr<boost::asio::ssl::context> on_tls_init(websocketpp::connection_hdl) {
        auto ctx = websocketpp::lib::make_shared<boost::asio::ssl::context>(boost::asio::ssl::context::tlsv12_client);
        ctx->set_options(boost::asio::ssl::context::default_workarounds |
                         boost::asio::ssl::context::no_sslv2 |
                         boost::asio::ssl::context::no_sslv3 |
                         boost::asio::ssl::context::single_dh_use);
        return ctx;
    }

    void on_open(websocketpp::connection_hdl hdl) {
        // Start thread sending ping every few seconds
        m_thread = websocketpp::lib::thread { &debug_client::ping, this, hdl };
    }

    void on_pong_timeout(websocketpp::connection_hdl hdl, std::string payload) {
        if (m_timeouts < 2) {
            m_timeouts++;
            return;
        }

        std::cout << "PONG TIMEOUT!" << std::endl;
        m_endpoint.close(hdl, websocketpp::close::status::normal, "consecutive onPongTimeouts");
    }

    void ping(websocketpp::connection_hdl hdl) {
        auto con = m_endpoint.get_con_from_hdl(hdl);
        while (con->get_state() == websocketpp::session::state::open) {
            std::cout << "Sending heartbeat" << std::endl;
            websocketpp::lib::error_code ec;
            m_endpoint.ping(hdl, "", ec);
            if (ec) {
                std::cout << "Ping error: " << ec << std::endl;
                break;
            }

            // Make sure this is longer than the pong timeout (currently 1000ms) so we
            // don't send another too quickly.
            std::this_thread::sleep_for(std::chrono::seconds(3));
        }
    }

    void wait() {
        m_thread.join();
    }
private:
    websocketpp::client<websocketpp::config::debug_asio_tls> m_endpoint;
    websocketpp::lib::thread m_thread;
    uint8_t m_timeouts = 0;
};

int main(int argc, char* argv[]) {
    // Matches echo_server_tls
    std::string uri = "wss://localhost:9002";

    if (argc == 2) {
        uri = argv[1];
    }

    debug_client endpoint;
    endpoint.start(uri);
    endpoint.wait();
}
```
running against examples/echo_server_tls in an Ubuntu 18.04 container. I'm using Boost 1.65, (although I've also encountered this with Boost 1.67).

If I connect that client to the example server, then introduce a firewall rule to drop outbound messages to the server, as in
```
iptables -A OUTPUT -p tcp --destination-port 9002 -j DROP; iptables -S
```
then immediately after the pong timeout triggers closing the connection (which prints `PONG TIMEOUT!`) and before the close timeout (of 5s) I remove the iptables rule, the server gets stuck in a busy loop after closing the connection.

Logs in that state look like
```
PONG TIMEOUT!
[2020-05-15 22:01:27] [devel] connection close
[2020-05-15 22:01:27] [devel] send_close_frame
[2020-05-15 22:01:27] [devel] closing with specified codes
[2020-05-15 22:01:27] [devel] Closing with code: 1000, and reason: consecutive onPongTimeouts
[2020-05-15 22:01:27] [devel] write_push: message count: 1 buffer size: 28
[2020-05-15 22:01:27] [devel] write_pop: message count: 0 buffer size: 0
[2020-05-15 22:01:27] [frame_header] Dispatching write containing 1 message(s) containing 6 header bytes and 28 payload bytes
[2020-05-15 22:01:27] [frame_header] Header Bytes:
[0] (6) 88 9C 00 00 00 00

[2020-05-15 22:01:27] [frame_payload] Payload Bytes:
[0] (28) [8] 03 E8 63 6F 6E 73 65 63 75 74 69 76 65 20 6F 6E 50 6F 6E 67 54 69 6D 65 6F 75 74 73

[2020-05-15 22:01:27] [devel] connection handle_write_frame
[2020-05-15 22:01:32] [devel] asio close handshake timer expired
[2020-05-15 22:01:32] [devel] connection terminate
[2020-05-15 22:01:32] [devel] asio connection async_shutdown
[2020-05-15 22:01:32] [devel] asio con handle_async_read
[2020-05-15 22:01:32] [devel] p = 0 bytes transferred = 2
[2020-05-15 22:01:32] [devel] calling consume with 2 bytes
[2020-05-15 22:01:32] [devel] Processing Bytes: 8A 00
[2020-05-15 22:01:32] [devel] bytes left after consume: 0
[2020-05-15 22:01:32] [devel] Complete message received. Dispatching
[2020-05-15 22:01:32] [devel] process_control_frame
[2020-05-15 22:01:32] [control] Control frame received with opcode 10
[2020-05-15 22:01:32] [warning] got frame in state closed
[2020-05-15 22:01:32] [devel] asio async_read_at_least: 1
[2020-05-15 22:01:32] [devel] asio con handle_async_read
[2020-05-15 22:01:32] [devel] p = 0 bytes transferred = 2
[2020-05-15 22:01:32] [devel] calling consume with 2 bytes
[2020-05-15 22:01:32] [devel] Processing Bytes: 8A 00
[2020-05-15 22:01:32] [devel] bytes left after consume: 0
[2020-05-15 22:01:32] [devel] Complete message received. Dispatching
[2020-05-15 22:01:32] [devel] process_control_frame
[2020-05-15 22:01:32] [control] Control frame received with opcode 10
[2020-05-15 22:01:32] [warning] got frame in state closed
[2020-05-15 22:01:32] [devel] asio async_read_at_least: 1
[2020-05-15 22:01:32] [info] asio async_shutdown error: asio.ssl:336150819 (application data after close notify)
[2020-05-15 22:01:32] [devel] connection handle_terminate
[2020-05-15 22:01:32] [devel] handle_terminate error: websocketpp.transport:11 (Generic TLS related error)
[2020-05-15 22:01:32] [disconnect] Disconnect close local:[1006,The closing handshake timed out] remote:[1006]
[2020-05-15 22:01:32] [devel] asio socket shutdown timer cancelled
[2020-05-15 22:01:32] [devel] asio con handle_async_read
[2020-05-15 22:01:32] [devel] p = 0 bytes transferred = 0
[2020-05-15 22:01:32] [devel] asio async_read_at_least: 1
[2020-05-15 22:01:32] [devel] asio con handle_async_read
```
The last 3 lines continue until I kill the process.

Sometimes that doesn't happen and the server terminates, which looks like
```
PONG TIMEOUT!
[2020-05-15 22:16:54] [devel] connection close
[2020-05-15 22:16:54] [devel] send_close_frame
[2020-05-15 22:16:54] [devel] closing with specified codes
[2020-05-15 22:16:54] [devel] Closing with code: 1000, and reason: consecutive onPongTimeouts
[2020-05-15 22:16:54] [devel] write_push: message count: 1 buffer size: 28
[2020-05-15 22:16:54] [devel] write_pop: message count: 0 buffer size: 0
[2020-05-15 22:16:54] [frame_header] Dispatching write containing 1 message(s) containing 6 header bytes and 28 payload bytes
[2020-05-15 22:16:54] [frame_header] Header Bytes:
[0] (6) 88 9C 00 00 00 00

[2020-05-15 22:16:54] [frame_payload] Payload Bytes:
[0] (28) [8] 03 E8 63 6F 6E 73 65 63 75 74 69 76 65 20 6F 6E 50 6F 6E 67 54 69 6D 65 6F 75 74 73

[2020-05-15 22:16:54] [devel] connection handle_write_frame
[2020-05-15 22:16:59] [devel] asio close handshake timer expired
[2020-05-15 22:16:59] [devel] connection terminate
[2020-05-15 22:16:59] [devel] asio connection async_shutdown
[2020-05-15 22:16:59] [devel] asio con handle_async_read
[2020-05-15 22:16:59] [devel] p = 0 bytes transferred = 2
[2020-05-15 22:16:59] [devel] calling consume with 2 bytes
[2020-05-15 22:16:59] [devel] Processing Bytes: 8A 00
[2020-05-15 22:16:59] [devel] bytes left after consume: 0
[2020-05-15 22:16:59] [devel] Complete message received. Dispatching
[2020-05-15 22:16:59] [devel] process_control_frame
[2020-05-15 22:16:59] [control] Control frame received with opcode 10
[2020-05-15 22:16:59] [warning] got frame in state closed
[2020-05-15 22:16:59] [devel] asio async_read_at_least: 1
[2020-05-15 22:16:59] [devel] asio con handle_async_read
[2020-05-15 22:16:59] [devel] p = 0 bytes transferred = 2
[2020-05-15 22:16:59] [devel] calling consume with 2 bytes
[2020-05-15 22:16:59] [devel] Processing Bytes: 8A 00
[2020-05-15 22:16:59] [devel] bytes left after consume: 0
[2020-05-15 22:16:59] [devel] Complete message received. Dispatching
[2020-05-15 22:16:59] [devel] process_control_frame
[2020-05-15 22:16:59] [control] Control frame received with opcode 10
[2020-05-15 22:16:59] [warning] got frame in state closed
[2020-05-15 22:16:59] [devel] asio async_read_at_least: 1
[2020-05-15 22:16:59] [devel] asio con handle_async_read
[2020-05-15 22:16:59] [devel] p = 0 bytes transferred = 2
[2020-05-15 22:16:59] [devel] calling consume with 2 bytes
[2020-05-15 22:16:59] [devel] Processing Bytes: 8A 00
[2020-05-15 22:16:59] [devel] bytes left after consume: 0
[2020-05-15 22:16:59] [devel] Complete message received. Dispatching
[2020-05-15 22:16:59] [devel] process_control_frame
[2020-05-15 22:16:59] [control] Control frame received with opcode 10
[2020-05-15 22:16:59] [warning] got frame in state closed
[2020-05-15 22:16:59] [devel] asio async_read_at_least: 1
[2020-05-15 22:16:59] [info] asio async_shutdown error: asio.misc:2 (End of file)
[2020-05-15 22:16:59] [devel] connection handle_terminate
[2020-05-15 22:16:59] [devel] handle_terminate error: websocketpp.transport:2 (Underlying Transport Error)
[2020-05-15 22:16:59] [disconnect] Disconnect close local:[1006,The closing handshake timed out] remote:[1006]
[2020-05-15 22:16:59] [devel] asio socket shutdown timer cancelled
[2020-05-15 22:16:59] [devel] asio con handle_async_read
[2020-05-15 22:16:59] [devel] got eof from closed con
```

Several tweaks seemed to important to reproducing this failure
- using `boost::asio::ssl::context::tlsv12_client` instead of `boost::asio::ssl::context::tlsv1`
- accumulating multiple (at least 3) dropped pings

**Comment by MikaelSmith** [comment]
Ok, looks like this is probably fixed by https://github.com/chriskohlhoff/asio/issues/261, which appears to be present in Boost 1.71. Still checking, but if that turns out to fix it I'll close the issue.

**Comment by MikaelSmith** [comment]
Ok, I'm fairly confident this is fixed by using Boost 1.71.
